python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+    # Start the workflow
+    handle = await client.start_workflow(GreetingWorkflow.run, "John")
+
+    # Confirm the current greeting
+    assert "Hello, John!" == await handle.query(GreetingWorkflow.current_greeting)
+
+    # Change the salutation part
+    await handle.signal(GreetingWorkflow.update_salutation, "Hola")
+
+    # Now tell it to complete with the new greeting
+    await handle.signal(GreetingWorkflow.complete_with_greeting)
+
+    # Check result
+    assert "Hola, John!" == await handle.result()
+
+def create_worker(client: Client) -> Worker:
+    return Worker(
+      client=client,
+      activities=[create_greeting_activity],
+      workflows=[GreetingWorkflow],
+    )
+```
+
+Notes:
+
+* Workflows are classes
+  * Consideration was given to making workflows be functions (in addition to classes or having the class equivalent sit
+    on top) with several possibilities for defining signals and queries, but all were too ugly/complicated
+* `@activity.defn`
+  * Must be on function and accepts an optional activity `name` which defaults to function `__name__`
+  * This was originally not wanted in the activity impl (phase 1) due to complaints of the name not being obvious
+  * We need to define the activity name on the activity for execution by the workflow
+  * Is it ok to now make this required? Is it ok to now make the name implicit? Is it ok to now use a list instead of a
+    dict to specify activities for the worker?
+* `@workflow.defn`
+  * Must be on class and accepts an optional workflow `name` which defaults to class `__name__`
+  * This is required to clearly mark the class as a workflow
+  * No base classes can have this value
+* `@workflow.run`
+  * Must be on a single async method (regardless of inheritance level) and accepts no params
+  * The params represent the workflow params and the response represents the workflow response
+    * Is it too annoying that the params aren't on `__init__` instead? The problem with being on a constructor is you
+      can't do async logic there so that's annoying too.
+  * An additional helper for `@workflow.start` was considered to allow "async complete" of workflow via another call but
+    it became too confusing
+  * The reference to this function is used for starting a workflow with a typed handle
+* `@workflow.signal`
+  * Can be on any number of methods (async or not, can be inherited)
+  * Can have an optional `name` which defaults to function `__name__`
+  * Params are signal params (only positional params allowed), cannot return value
+  * Can have optional `dynamic` boolean which, if set to true, means the signal is a catch-all for otherwise unhandled
+    signals. `name` cannot be present and there must be two params: a name param and a varargs param
+    * An alternative approach that is similarly Pythonic would be to have a `__temporal_signal__` function that could be
+      defined to catch them, but the decorator is clearer
```

Comment:
The special function has the benefit that there can only be one, which should be the case. It's uglier, though. Probably we just want to enforce at init time that you've only defined one catch-all handler.

Reply:
Agreed